designs/sam-debugging/local-sam-debugging.md

```diff
@@ -0,0 +1,336 @@
+
+1.  the debug configuration is validated (see [Debug Configuration Validations](#debug-configuration-validation)). The launch is stopped if errors are detected.
+1.  a SAM Application's code is built
+1.  a SAM Template resource is invoked
+1.  a debugger is attached to the invoked resource (this is skipped if the debug configuration was launched using "Run without Debugging")
+    -   debug output is shown in the Debug Console of VS Code during this Debug Session
+
+The Debug Console can only be written to when VS Code has an active debug session. During the launch sequence, launch progress is written to the Toolkit's Output Channel until a debugger is attached. SAM CLI commands executed by the Toolkit are written to the Output Channel, so users can see what is happening behind the scenes.
+
+Debug configurations reside in a JSON file. The Toolkit assists users working with `aws-sam` entries in the following ways:
+
+-   users see autocompletion for `aws-sam` related fields
+    -   autocompletion is not available for open ended configuration values. For example, when a user types in the location of a SAM Template file, there is no filesystem-based autocompletion. The toolkit validates `aws-sam` debug configurations when launched, and notifies users as errant values are detected.
+-   users see field descriptions (tooltips) for `aws-sam` related fields
+-   users have access to snippets that produce typical (or starter) `aws-sam` debug configurations
+-   the toolkit is capable of generating an `aws-sam` Debug Configuration for all `AWS::Serverless::Function` and `AWS::Lambda::Function` resources detected within a workspace
+    -   when users create a new SAM Application, the toolkit automatically performs this step to produce initial Debug Configurations
+
+Example Debug Configuration entries can be found in the [Appendix](#sample-debug-configurations).
+
+### <a id="codelenses"></a> CodeLenses
+
+The Toolkit uses [CodeLenses](#terms-codelenses) as a way of setting up a [Debug Configuration](#debug-configurations) of type `aws-sam`.
+
+Some users find CodeLenses visually distracting. Other users use the Toolkit for features that aren't related to local debugging. The Toolkit's CodeLenses are enabled by default, but can be disabled in the Toolkit settings.
+
+The CodeLenses discussed below only appear for languages/runtimes that the Toolkit provides support for (see [Introduction](#introduction)).
+
+#### CodeLenses in SAM Template files
+
+The Toolkit offers a way for users to set up debug configurations for template resources of type `AWS::Serverless::Function` and `AWS::Lambda::Function`.
+
+The Toolkit adds an "Add Debug Configuration" CodeLens above resources that are not already referenced by a debug configuration. When users click this CodeLens, the toolkit produces a Debug Configuration of type `aws-sam`, configures it to invoke the resource, and adds it to `launch.json`. Users are taken to the new entry in `launch.json`.
```

Comment:
Let's respecify as "Add Launch Configuration" as 1: technically that's what they're called and 2: VS Code is de-emphasizing the "Debug" functionality in favor of "Run": https://code.visualstudio.com/updates/v1_43#_run-more-prominent-in-the-ui